Dpu test plan phase 2

nissampa opened the pull request:
Added Phase-2 test plan for Smartswitch platform tests

Check memory on DPU
Check DPU status and pcie Link after memory exhaustion on Switch
Check DPU status and pcie Link after memory exhaustion on DPU
Check DPU status and pcie Link after restart pmon
Check DPU status and pcie Link after reload of configuration
Check DPU status and pcie Link after kernel panic on Switch
Check DPU status and pcie Link after kernel panic on DPU
Check DPU status and pcie Link after switch power cycle
Check DPU status and pcie Link after SW trip by temperature trigger

vvolam commented:
> Check memory on DPU
> Check DPU status and pcie Link after memory exhaustion on Switch
> Check DPU status and pcie Link after memory exhaustion on DPU
> Check DPU status and pcie Link after restart pmon
> Check DPU status and pcie Link after reload of configuration
> Check DPU status and pcie Link after kernel panic on Switch
> Check DPU status and pcie Link after kernel panic on DPU
> Check DPU status and pcie Link after power off reboot
> Check DPU status and pcie Link after SW trip by temperature trigger

I think following additional condition also can be considered.

"Check DPU status and PCIe Link after network interface flapping"

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,672 @@ root@sonic:/home/cisco# show chassis modules status
+#### Pass/Fail Criteria
+
+ * Verify that used memory should not cross the specified threshold value (90) of total memory.
+ * Threshold can be set different based on platform.
```
Where can this threshold be set? will it be REDIS database or platform.json?

nissampa replied:
It can be in platform.json.

vvolam replied:
Do we have this change or PR to set the threshold?

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,672 @@ root@sonic:/home/cisco# show chassis modules status
+
+   * Use `sudo swapoff -a`. Swapping is turned off so the OOM is triggered in a shorter time.
+   * Use 'nohup bash -c "sleep 5 && tail /dev/zero" &' to to run out of memory completely.
+   * It runs on the background and `nohup` is also necessary to protect thebackground process.
```
nit: space needed at "thebackground"

nissampa replied:
modified.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,672 @@ root@sonic:/home/cisco# show chassis modules status
+   * Added `sleep 5` to ensure ansible receive the result first.
+   * Check the status and power on DPUs after switch goes for reboot and comes back
+   * Use `show chassis modules status` to check status of the DPUs.
+   * Append to the existing test case: https://github.com/sonic-net/sonic-mgmt/blob/master/tests/platform_tests/test_memory_exhaustion.py
```
Can we have a negative test case, where DPU cannot recover gracefully and a hardware reset comes into play?

nissampa replied:
Even this cli does kind of power cycling rather than graceful reboot.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,672 @@ root@sonic:/home/cisco# show chassis modules status
+   * Added `sleep 5` to ensure ansible receive the result first.
+   
+    - Switch:
+       * Powercycling of DPU is to ensure that pcie link came up properly after the memory exhaustion test.
```
In case of test failure for all these PCIe Link check, can we make sure of add some diagnostics logs for easy troubleshooting? Either you can mention here or can be added when writing test case.

nissampa replied:
Yes, I will make sure to have logs included in test scripts.

nissampa commented:
> > Check memory on DPU
> > Check DPU status and pcie Link after memory exhaustion on Switch
> > Check DPU status and pcie Link after memory exhaustion on DPU
> > Check DPU status and pcie Link after restart pmon
> > Check DPU status and pcie Link after reload of configuration
> > Check DPU status and pcie Link after kernel panic on Switch
> > Check DPU status and pcie Link after kernel panic on DPU
> > Check DPU status and pcie Link after power off reboot
> > Check DPU status and pcie Link after SW trip by temperature trigger
> 
> I think following additional condition also can be considered.
> 
> "Check DPU status and PCIe Link after network interface flapping"

It can be added as HA case as we discussed in our community call.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -11,12 +12,21 @@
+    - [1.17 Check DPU status and pcie Link after reload of configuration](#117-check-dpu-status-and-pcie-link-after-reload-of-configuration)
+    - [1.18 Check DPU status and pcie Link after kernel panic on Switch](#118-check-dpu-status-and-pcie-link-after-kernel-panic-on-switch)
+    - [1.19 Check DPU status and pcie Link after kernel panic on DPU](#119-check-dpu-status-and-pcie-link-after-kernel-panic-on-dpu)
+    - [1.20 Check DPU status and pcie Link after power off reboot](#120-check-dpu-status-and-pcie-link-after-power-off-reboot)
```
Instead, could you rephrase this way for clarity "1.20 Check DPU status and pcie Link after power cycle DPUs?"

nissampa replied:
Sure, Changed it.
This is for switch power cycle only.

nissampa replied:
Changed it and committed

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -11,12 +12,21 @@
+    - [1.17 Check DPU status and pcie Link after reload of configuration](#117-check-dpu-status-and-pcie-link-after-reload-of-configuration)
+    - [1.18 Check DPU status and pcie Link after kernel panic on Switch](#118-check-dpu-status-and-pcie-link-after-kernel-panic-on-switch)
+    - [1.19 Check DPU status and pcie Link after kernel panic on DPU](#119-check-dpu-status-and-pcie-link-after-kernel-panic-on-dpu)
+    - [1.20 Check DPU status and pcie Link after power off reboot](#120-check-dpu-status-and-pcie-link-after-power-off-reboot)
```
Also, do we have a test case to check DPU status and PCIe link after Switch power cycle

nissampa replied:
Changed it and committed.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,673 @@ root@sonic:/home/cisco# show chassis modules status
+#### Steps
+
+ * Infrastructure will be provided to run the scripts that triggers the temperature trip based on vendor.
+ * The following is the example sequence to trigger temperature trip on the dpu
```
Do we have any API which can call vendor specific implementation?

nissampa replied:
It will be implemented as part of test scripts. The plan is to have function that takes vendor specific scripts to be run and checks the output. 

There is not existing api implementation.

nissampa commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -404,9 +431,9 @@ On Switch: Showing example of one DPU pcie link
 root@sonic:/home/cisco# show platform pcieinfo -c
 
-root@sonic:/home/cisco# echo 1 > /sys/bus/pci/devices/0000:1a:00.0/remove
+root@sonic:/home/cisco# config chassis modules shutdown DPU<DPU_NUM>
```
This can be considered as both, since while shutting down it detach the pcie.
While powering up it does rescan. That is why we dont need to have dpu power cycle case separately.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,673 @@ root@sonic:/home/cisco# show chassis modules status
+ * Use command `show system-memory` to get memory usage on each of those DPUs
+ * Use `show system-health dpu <DPU_NUM>` to check memory check service status
+ * Use `pdsctl show system --events` to check memory related events triggered.
+   (This is vendor specific event monitoring cli) 
```
Does it mean that the command is specific to Cisco? or does it invoke vendor specific API?

nissampa replied:
Yes last command is vendor specific. 
The objective is to provide general function which expects vendor specific cli and expected output. 
Then the function just can execute and compare it to provide the result

vvolam replied:
Is so, can you capture NVIDIA specific command as well?

nissampa replied:
I am not sure about Nvidia's command. 
But during the call, they agreed to it on first two steps itself. The third one I added it extra check from our side. 
If any one wants to add any extra checks, infra will be provided.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,673 @@ root@sonic:/home/cisco# show chassis modules status
+
+ * Infrastructure support will be provided to choose from following clis or in combinations based on the vendor.
+ * Use command `show system-memory` to get memory usage on each of those DPUs
+ * Use `show system-health dpu <DPU_NUM>` to check memory check service status
```
Is this command to be run on DPU or NPU? If it is NPU, could you clarify it here?

vvolam replied:
Also, when you have mix of commands to run on NPU and DPU, could you please clarify where you are running the command in the text in all the tests?

nissampa replied:
It is on DPU. 

Yes I understand that it is confusing as we jump between npu and dpu. 
I tried to mention that to an extent on each cases on where it should be running. 

- I have the section called verify in to tell explicitly on where to run in all the cases. 
- In sample output as well, I mentioned before on where it is running. 
- On title also for all dpu run cases, I have titled as on dpu.

vvolam replied:
As some of the tests are having commands to check in both NPU and DPU. Could you add "NPU:" or "DPU:" prefix before mentioning the command to use? Will it be more clear?

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -36,6 +46,14 @@ Purpose of the test is to verify smartswich platform related functionalities/fea
+## Testbed and Version
+
+The test runs on the os versions 2024011 and above.
+Check is to be in place for that and skips if it is below the mentioned version.
```
Could you rephrase with "Add a check to confirm that the test environment uses version 2024011 or later; if the version is earlier, skip the test." or something similar?

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -384,7 +411,7 @@ root@sonic:/home/cisco# show chassis modules status
 
 
-### 1.7 Check removal of pcie link between NPU and DPU
+### 1.7 Check pcie link status between NPU and DPU
```
Nit: Optional, using "Verify" in place of "Check" in each of the test header provides most consistent wording.

nissampa replied:
Actually, I consistently followed "Check" in all the test cases.

Can we leave it as such to maintain it?

vvolam replied:
okay.

nissampa commented:
@vvolam 
Do we have this change or PR to set the threshold?

To comment on the last question, it will be handled in test scripts.

nissampa commented:
@vvolam, Can we close on this? if things are good.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -36,6 +46,14 @@ Purpose of the test is to verify smartswich platform related functionalities/fea
 
+## Testbed and Version
+
+The test runs on the os versions 2024011 and above.
```
Is it 202411?

nissampa replied:
Changed it.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -56,12 +74,21 @@ General convention of DPU0, DPU1, DPU2 and DPUX has been followed to represent D
 | 1.9 | Check the State of DPUs      | To Verify DPU state details during online and offline      | |
 | 1.10 | Check the Health of DPUs       | To Verify overall health (LED, process, docker, services and hw) of DPU | Phase:2 |
 | 1.11 | Check reboot cause history       | To Verify reboot cause history cli | |
 | 1.12 | Check the DPU state after OS reboot       | To Verify DPU state on host reboot | |
```
Is this Switch reboot or DPU OS reboot, could you update the sentence accordingly.

nissampa replied:
Changed it to say switch reboot

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,666 @@ root@sonic:/home/cisco# show chassis modules status
+                             dpu_control_plane_state  UP             20241003 17:57:22  All containers are up and running, host-ethlink-status: Uplink1/1 is UP
+                             dpu_data_plane_state     UP             20241001 19:54:30  DPU container named polaris is running, pdsagent running : OK, pciemgrd running : OK
+root@MtFuji:/home/cisco# 
+
```
Could you add a comment here about the HW sku that is being executed?

nissampa commented:
@vvolam 
- Made changes to use switch consistently across the document instead of NPU
- Made changes to say use command on switch/dpu based on the commands in the steps.
- Made changes to 202411 instead of 2024011
- Made other changes as we discussed on the call.

nissampa commented:
> > @vvolam
> > 
> > * Made changes to use switch consistently across the document instead of NPU
> > * Made changes to say use command on switch/dpu based on the commands in the steps.
> > * Made changes to 202411 instead of 2024011
> > * Made other changes as we discussed on the call.
> 
> @nissampa forgot to mention in the call, could you add topology information section also, mentioning new topo added for smartswitch and also all t1 cases should be running on smartswitch?

Added it in new section called topology mentioning both the points.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +672,667 @@ root@sonic:/home/cisco# show chassis modules status
+Swap:              0           0           0
+root@sonic:/home/admin# 
+root@sonic:/home/admin# 
+root@MtFuji:/home/cisco# show system-health dpu DPU0
```
It may not be related here but interesting why do we need to provide dpu module_name when we are running the command on the DPU?

nissampa replied:
Sorry, this one command on switch.

I changed it.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +678,667 @@ root@sonic:/home/cisco# show chassis modules status
+                             dpu_data_plane_state     UP             20241001 19:54:30  DPU container named polaris is running, pdsagent running : OK, pciemgrd running : OK
+root@MtFuji:/home/cisco# 
+
+# Note: This command is run on cisco specific testbed.
```
could you mention the HWSKU of the testbed here?

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -36,6 +47,19 @@ Purpose of the test is to verify smartswich platform related functionalities/fea
+## Topology
+
+New topology called smartswitch-t1 has been added for running smartswitch cases.
+T1 cases also runs on the new topology. 
```
Do we run 'any' topo test cases also on smartswitch?

nissampa replied:
Yes it should.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +678,667 @@ root@sonic:/home/cisco# show chassis modules status
+#### Steps
+
+* Use command on DPU: `sudo swapoff -a`. Swapping is turned off so the OOM is triggered in a shorter time.
+* Use command on DPU: 'nohup bash -c "sleep 5 && tail /dev/zero" &' to to run out of memory completely.
```
nit: "to" is repeated twice

nissampa replied:
Changed it.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +678,667 @@ root@sonic:/home/cisco# show chassis modules status
+
+#### Steps
+* Use command on DPU: `nohup bash -c "sleep 5 && echo c > /proc/sysrq-trigger" &`.
+* Powercycling of DPU is to ensure that pcie link came up properly after the memory exhaustion test.
```
Do you mean kernel panic on DPU here? as there is a reboot type for KERNEL_PANIC instead of memory exhaustion?

nissampa replied:
Yes, its on DPU.

I added it similar to what is happening on switch. reboot_type kernel panic does like this underneath on the code that existed in sonic-mgmt. So, added the command directly.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +678,667 @@ root@sonic:/home/cisco# show chassis modules status
+#### Pass/Fail Criteria
+ * Verify number of DPUs from inventory file for the testbed and number of DPUs shown in the cli output.
+ * Verify Ping works to all the mid plane ip listed in the ansible inventory file for the testbed.
+ * Verify `show reboot-cause <DPU_Number>` to check the reboot is caused by kernel panic. 
```
Is this verified on DPU or Switch? could you prefix please?

nissampa replied:
Its on switch.
Added the prefix
Not tested as it still in implementation

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -645,6 +678,667 @@ root@sonic:/home/cisco# show chassis modules status
+
+ * Infrastructure will be provided to run the scripts that triggers the temperature trip based on vendor.
+ * The following is the example sequence to trigger temperature trip on the DPU
+     - Note: If Cisco setup, the following steps work.
```
Please provide HWSKU information of cisco example used here.

nissampa replied:
Added it.

vvolam reviewed (commented):
@nissampa some minor comments. otherwise lgtm. Thanks for addressing all the comments.

nissampa commented:
> @nissampa some minor comments. otherwise lgtm. Thanks for addressing all the comments.

Thank you. Addressed minor comments as well. 
If all looks good, Can you help me approve and merge it ?

nissampa commented:
@vvolam, Let me know if this test plan PR is good and can be given approval.

vvolam commented on docs/testplan/Smartswitch-test-plan.md:
```diff
@@ -1336,7 +1344,7 @@ root@sonic:/home/cisco# show reboot-cause history dpu0
 #### Pass/Fail Criteria
  * Verify Ping works to DPU  mid plane ip listed in the ansible inventory file for the testbed.
- * Verify show reboot cause history <dpu> to check the cause as cattrip.
+ * Verify the command on Swithc: show reboot cause history <dpu> to check the cause as cattrip.
```
nit: Switch typo.

vvolam reviewed (approved):
Other than typo, looks good to me. Thank you!
